maintenance/testNotification.php

```diff
@@ -0,0 +1,167 @@
+		$this->output( "Adding notification '$notificationKey'\n" );
+		$this->output( "Agent is '{$this->agentUser->getName()}' (ID:{$this->agentUser->getId()})\n" );
+		$this->output( "Subject title is '{$this->title->getPrefixedDbKey()}' (ID:{$this->title->getArticleId()})\n" );
+
+		BSNotifications::notify(
+			$notificationKey,
+			$this->agentUser,
+			$this->title,
+			$this->makeExtraParams()
+		);
+
+		$this->output( "Done.\n" );
+	}
+
+	protected function makeAgentUser() {
+		$agent = $this->getOption( 'agent', '' );
+		if( !empty( $agent ) ) {
+			$this->agentUser = User::newFromName( $agent );
+			if( $this->agentUser instanceof User === false || $this->agentUser->getId() === 0 ) {
+				throw  new Exception( "Invalid or not existing user: $agent" );
+			}
+		}
+		else {
+			$this->agentUser = User::newFromId( 1 );
+		}
+	}
+
+	protected function makeTitle() {
+		$title = $this->getOption( 'title', '' );
+		if( !empty( $title ) ) {
+			$this->title = Title::newFromText( $title );
+		}
+		else {
+			$this->title = Title::newMainPage();
+		}
+	}
+
+	protected function makeExtraParams() {
+		$aExtra = [];
+		$aAffectedUsers = $this->getAffectedUsers();
+		if( !empty( $aAffectedUsers ) ) {
+			$aExtra['affected-users'] = $aAffectedUsers;
+		}
+		$aAffectedGroups = $this->getAffectedGroups();
+		if( !empty( $aAffectedGroups ) ) {
+			$aExtra['affected-groups'] = $aAffectedGroups;
+		}
+
+		return $aExtra;
+	}
+
+	protected function setupAlternateUserMailer() {
+		if( !$this->getOption(  'outputMail' , false ) ) {
+			return;
+		}
+
+		Hooks::register( 'AlternateUserMailer', function( $headers, $to, $from, $subject, $body ) {
+			$this->outputMail( $headers, $to, $from, $subject, $body );
+			return false;
+		});
+
+		//Do a test
+		UserMailer::send( 
+			[ new MailAddress( 'support@hallowelt.com') ],
+			new MailAddress( 'info@hallowelt.com' ),
+			'Hello World',
+			'Lorem ipsum dolor sit amet'
+		);
+	}
+
+	/**
+	 *
+	 * @param array $headers
+	 * @param MailAddress[] $to
+	 * @param MailAddress $from
+	 * @param string $subject
+	 * @param string $body
+	 */
+	protected function outputMail( $headers, $to, $from, $subject, $body ) {
+		$out = [];
+		$out[] = '###############';
+		foreach( $headers as $headerName => $headerValue ) {
+			$out[] = "$headerName: $headerValue";
+		}
+
+		$out[] = '---------------------------------------------';
+		$out[] = 'FROM: '.$from->toString();
+
+		$tos = [];
+		foreach( $to as $mailAdress ) {
+			$tos[] = $mailAdress->toString();
+		}
+		$out[] = 'TO: '.implode( '; ', $tos );
+		$out[] = 'SUBJECT: '.$subject;
+		$out[] = 'BODY:';
+		$out[] = $body;
+		$out[] = '';
+
+		$this->output( implode( "\n", $out) );
+	}
+
+	protected function getAffectedUsers() {
+		$sAffectedUsers = $this->getOption( 'affectedusers', '' );
+		$aUserNames = explode( ',', $sAffectedUsers );
+		$aUsers = [];
+		foreach( $aUserNames as $sUserName ) {
+			$aUsers[] = User::newFromName( $sUserName );
```

Comment:
I belive "affected users" arrays should be an array of ints, not User objects, I might be wrong

Reply:
No, User objects are just fine: https://github.com/hallowelt/mediawiki-extensions-BlueSpiceEchoConnector/blob/master/includes/EchoConnectorHooks.php#L172